main/src/library/HashSet.flix

```diff
@@ -0,0 +1,546 @@
+/*
+ * Copyright 2017 Liam Palmer, Esben Bjerre
+ * Copyright 2024 Jonathan Lindegaard Starup
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+///
+/// The HashSet type.
+///
+/// A set is currently represented internally as a red-black tree.
+///
+pub enum HashSet[t] with Sendable, Eq {
+    case HashSet(RedBlackTree[Int32, HashSet.ListSet[t]])
+}
+
+instance ToString[HashSet[t]] with ToString[t] {
+    pub def toString(x: HashSet[t]): String = HashSet.toString(x)
+}
+
+instance Collectable[HashSet[a]] with Eq[a], Hash[a] {
+    type Elm = a
+    pub def collect(iter: Iterator[a, ef, r]): HashSet[a] \ { ef, r } =
+        HashSet.collectI(iter)
+}
+
+instance Hash[HashSet[t]] with Hash[t] {
+    pub def hash(s: HashSet[t]): Int32 =
+        HashSet.foldLeft((acc, x) -> acc `Hash.combine` Hash.hash(x), Hash.magic(), s)
+}
+
+instance SemiGroup[HashSet[a]] with Eq[a], Hash[a] {
+    pub def combine(x: HashSet[a], y: HashSet[a]): HashSet[a] = HashSet.union(x,y)
+}
+
+instance Monoid[HashSet[a]] with Eq[a], Hash[a] {
+    pub def empty(): HashSet[a] = HashSet.empty()
+}
+
+instance Foldable[HashSet] {
+    pub def foldLeft(f: (b, a) -> b \ ef, s: b, s1: HashSet[a]): b \ ef = HashSet.foldLeft(f, s, s1)
+    pub def foldRight(f: (a, b) -> b \ ef, s: b, s1: HashSet[a]): b \ ef = HashSet.foldRight(f, s, s1)
+    pub def foldRightWithCont(f: (a, Unit -> b \ ef) -> b \ ef, s: b, s1: HashSet[a]): b \ ef = HashSet.foldRightWithCont(f, s, s1)
+}
+
+instance Iterable[HashSet[a]] {
+    type Elm = a
+    pub def iterator(rc: Region[r], s: HashSet[a]): Iterator[a, r, r] \ r = HashSet.iterator(rc, s)
+}
+
+mod RedBlackTree {
+    /// removes `k` if `f` return `None`
+    pub def updateFilter(f: (k, v) -> Option[v] \ ef, k: k, t: RedBlackTree[k, v]): RedBlackTree[k, v] \ ef with Order[k] = {
+        match get(k, t) {
+            case Some(v) => match f(k, v) {
+                case None => RedBlackTree.remove(k, t)
+                case Some(v1) => RedBlackTree.insert(k, v1, t)
+            }
+            case None => t
+        }
+    }
+}
+
+mod HashSet {
+    use ListSet.liftT
+    use ListSet.lift
+
+    /// Invariant: the list has no duplicates
+    pub enum ListSet[t](List[t])
+
+    instance Eq[ListSet[t]] with Eq[t] {
+        pub def eq(x: ListSet[t], y: ListSet[t]): Bool = ListSet.eq(x, y)
+    }
+
+    mod ListSet {
```

Comment:
Should ListSet be its own thing?

Reply:
It could, but its not very useful outside of this i'd think

HashMap needs ListMap too